src/extension.ts

```diff
@@ -1,97 +1,61 @@
+import { dirname, basename } from 'node:path'
 import * as vscode from 'vscode'
-import ansiEscapes from 'ansi-escapes'
 import { getExtensionSetting, registerExtensionCommand } from 'vscode-framework'
-import { jsLangs } from './util'
+import { parseVariables } from './utils'
 
 export const activate = () => {
-    // registerExtensionCommand('runWithSideCode', async () => {
-    //     const activeEditor = vscode.window.activeTextEditor
-    //     if (!activeEditor || activeEditor.viewColumn === undefined) return
-    //     await vscode.window.showTextDocument(vscode.Uri.parse(`${documentScheme}:${activeEditor.document.uri.toString(true)}`), {
-    //         preview: false,
-    //         viewColumn: vscode.ViewColumn.Beside,
-    //     })
-    // })
     type FsPath = string
     const activeTerminals = new Map<FsPath, vscode.Terminal>()
 
-    const checkRunButton = (textEditor: vscode.TextEditor | undefined): void => {
-        if (!textEditor || textEditor.viewColumn === undefined) return
-        let hasExec = !!getExtensionSetting('defaultExec')
-        if (!hasExec) hasExec = !!getExec(textEditor.document.languageId)
-        void vscode.commands.executeCommand('setContext', `terminal-code-runner.runBututon`, hasExec)
+    const checkDisplayRunButton = (textEditor: vscode.TextEditor | undefined): void => {
+        void vscode.commands.executeCommand('setContext', `terminal-code-runner.runButton`, getHasExec(textEditor))
     }
 
-    vscode.window.onDidChangeActiveTextEditor(checkRunButton)
-    checkRunButton(vscode.window.activeTextEditor)
+    vscode.window.onDidChangeActiveTextEditor(checkDisplayRunButton)
+    checkDisplayRunButton(vscode.window.activeTextEditor)
 
     registerExtensionCommand('runFile', async () => {
         const activeEditor = vscode.window.activeTextEditor
         if (!activeEditor || activeEditor.viewColumn === undefined) return
+
+        let exec = getExec(activeEditor)
+        if (!exec) {
+            void vscode.window.showWarningMessage(`No matched exec command!`)
+            return
+        }
+
         const { document } = activeEditor
         const { fsPath } = document.uri
         const fileDir = dirname(fsPath)
         const fileName = basename(fsPath)
-        const workspaceRoot = vscode.workspace.getWorkspaceFolder(document.uri)
-        const exec = (getExec(document.languageId) ?? getExtensionSetting('defaultExec'))
-            ?.replace(/\$workspaceRoot/g, workspaceRoot?.uri.fsPath ?? '')
-            .replace(/\$fileName/, fileName)
-            .replace(/\$dir/, fileDir)
-            .replace(/\$path/, fsPath)
-        if (!exec) {
-            void vscode.window.showWarningMessage(`No exec command for language ${document.languageId} is set`)
-            return
-        }
+        exec = await parseVariables(exec, document.uri)
 
-        const writeEmitter = new vscode.EventEmitter<string>()
-        const line = ''
+        // Uses different terminals for each file
```

Comment:
This though, I'm just curios, @tjx666 do you think it is useful? Eg shouldn't we reuse the same terminal instead (I mean add a setting to not spam with a lot of terminals when running different files)?

Reply:
added, and I set default false